include/fluent-bit/aws/flb_aws_imds.h

```diff
@@ -0,0 +1,109 @@
+/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*  Fluent Bit
+ *  ==========
+ *  Copyright (C) 2019-2021 The Fluent Bit Authors
+ *  Copyright (C) 2015-2018 Treasure Data Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#ifndef FLB_AWS_IMDS
+#define FLB_AWS_IMDS
+
+#define FLB_AWS_IMDS_HOST                                 "169.254.169.254"
+#define FLB_AWS_IMDS_HOST_LEN                             15
+#define FLB_AWS_IMDS_PORT                                 80
+
+#define FLB_AWS_IMDS_VERSION_EVALUATE                     0
+#define FLB_AWS_IMDS_VERSION_1                            1
+#define FLB_AWS_IMDS_VERSION_2                            2
+
+/* The following metadata paths can be evaluated with flb_aws_imds_request
+ * to retrieve specific metadata members */
+#define FLB_AWS_IMDS_INSTANCE_ID_PATH                       "/latest/meta-data/instance-id/"
+#define FLB_AWS_IMDS_AZ_PATH                                "/latest/meta-data/placement/availability-zone/"
+#define FLB_AWS_IMDS_INSTANCE_TYPE_PATH                     "/latest/meta-data/instance-type/"
+#define FLB_AWS_IMDS_PRIVATE_IP_PATH                        "/latest/meta-data/local-ipv4/"
+#define FLB_AWS_IMDS_VPC_ID_PATH_PREFIX                     "/latest/meta-data/network/interfaces/macs/"
+#define FLB_AWS_IMDS_AMI_ID_PATH                            "/latest/meta-data/ami-id/"
+#define FLB_AWS_IMDS_ACCOUNT_ID_PATH                        "/latest/dynamic/instance-identity/document/"
+#define FLB_AWS_IMDS_HOSTNAME_PATH                          "/latest/meta-data/hostname/"
+#define FLB_AWS_IMDS_MAC_PATH                               "/latest/meta-data/mac/"
+
+/* IMDS config values */
+struct flb_aws_imds_config {
+    int use_imds_version;  // FLB_AWS_IMDS_VERSION_EVALUATE for automatic detection
+};
+
```

Comment:
So I like the constants, esp FLB_AWS_IMDS_VERSION_EVALUATE, and I am fine with this separate config structure, but I want to say tho that it feels a bit like over-kill to have a structure for a single option.

Reply:
## Motivation
Here is my motivation behind the config structure. A default config structure is declared in imds.c. This default config structure is used as a config template when our IMDS struct is initialized. That way in the future, if we want to add some more options to customize the IMDS' behavior (e.g. fallback behavior or number of retries) we can add a default configuration in one place and not have to worry about changing IMDS initialization code in other places.

## Use
You can see the use of the default config structure used as a configuration template in create_ec2_provider() (and anywhere else IMDS is initialized future).

## Options
If you think that this ease of maintenance motivation is not strong enough to warrant the additional code, I am happy to explore other options.